core/src/main/java/com/scalar/db/storage/rpc/GrpcScanOnBidirectionalStream.java

```diff
@@ -100,31 +84,75 @@ private void throwIfError(ResponseOrError responseOrError) throws ExecutionExcep
     throwIfScannerHasNoMoreResults();
     ResponseOrError responseOrError =
         sendRequest(ScanRequest.newBuilder().setScan(ProtoUtil.toScan(scan)).build());
-    throwIfError(responseOrError);
+    throwIfErrorForOpenScanner(responseOrError);
     return getResults(responseOrError.getResponse());
   }
 
+  private void throwIfErrorForOpenScanner(ResponseOrError responseOrError)
+      throws ExecutionException {
+    if (responseOrError.isError()) {
+      hasMoreResults.set(false);
+      Throwable error = responseOrError.getError();
+      if (error instanceof StatusRuntimeException) {
+        StatusRuntimeException e = (StatusRuntimeException) error;
+        if (e.getStatus().getCode() == Code.INVALID_ARGUMENT) {
+          throw new IllegalArgumentException(e.getMessage());
+        }
+        if (e.getStatus().getCode() == Code.UNAVAILABLE) {
+          throw new ServiceTemporaryUnavailableException(e.getMessage());
+        }
+        throw new ExecutionException("failed to open scanner: " + e.getMessage());
+      }
+      if (error instanceof Error) {
+        throw (Error) error;
+      }
+      throw new ExecutionException("failed to open scanner", error);
+    }
+  }
+
   public List<Result> next() throws ExecutionException {
     throwIfScannerHasNoMoreResults();
     ResponseOrError responseOrError = sendRequest(ScanRequest.getDefaultInstance());
-    throwIfError(responseOrError);
+    throwIfErrorForNext(responseOrError);
     return getResults(responseOrError.getResponse());
   }
 
   public List<Result> next(int fetchCount) throws ExecutionException {
     throwIfScannerHasNoMoreResults();
     ResponseOrError responseOrError =
         sendRequest(ScanRequest.newBuilder().setFetchCount(fetchCount).build());
-    throwIfError(responseOrError);
+    throwIfErrorForNext(responseOrError);
     return getResults(responseOrError.getResponse());
   }
 
-  public void closeScanner() {
-    if (!hasMoreResults.get()) {
-      return;
+  private void throwIfErrorForNext(ResponseOrError responseOrError) throws ExecutionException {
+    if (responseOrError.isError()) {
+      hasMoreResults.set(false);
+      Throwable error = responseOrError.getError();
+      if (error instanceof StatusRuntimeException) {
+        StatusRuntimeException e = (StatusRuntimeException) error;
+        if (e.getStatus().getCode() == Code.INVALID_ARGUMENT) {
+          throw new IllegalArgumentException(e.getMessage());
+        }
```

Comment:
OK, that makes sense.
So, `throwIfErrorForOpenScanner` would also get UNAVAILABLE from other than pause?

I also feel it would be great if we could remove the redundancy.